File: internal/bench/servers/go/shared/io.go
// The io scenario's clients for both Go subjects: pgx for Postgres and go-redis
// for Redis, each pooled to the size every subject in the suite is pinned to.
//
// Opened only when the harness passes both URLs, so the other four scenarios run
// a process that has opened no socket and paid no connect in its startup number.
package shared

import (
	"context"
	"encoding/json"
	"net/http"
	"os"
	"strconv"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/redis/go-redis/v9"
)

const (
	ioRedisKey = "bench:greeting"
	ioRowID    = 1
	ioPoolSize = 8
	ioSelect   = "SELECT id, memo, amount FROM bench_ledger WHERE id = $1"
)

// IoPayload's field order is the JSON field order, and the harness compares bytes.
type IoPayload struct {
	Cached string `json:"cached"`
	ID     int32  `json:"id"`
	Memo   string `json:"memo"`
	Amount int32  `json:"amount"`
}

type Io struct {
	pool  *pgxpool.Pool
	redis *redis.Client
}

// NewIo returns nil when the harness did not enable the scenario, which is what
// the callers branch on before registering the route.
func NewIo(ctx context.Context) (*Io, error) {
	pgURL := os.Getenv("BENCH_IO_PG_URL")
	redisURL := os.Getenv("BENCH_IO_REDIS_URL")
	if pgURL == "" || redisURL == "" {
		return nil, nil
	}

	config, err := pgxpool.ParseConfig(pgURL)
	if err != nil {
		return nil, err
	}
	config.MaxConns = ioPoolSize
	pool, err := pgxpool.NewWithConfig(ctx, config)
	if err != nil {
		return nil, err
	}

	options, err := redis.ParseURL(redisURL)
	if err != nil {
		return nil, err
	}
	options.PoolSize = ioPoolSize
	client := redis.NewClient(options)

	io := &Io{pool: pool, redis: client}
	// Both clients connect lazily. Doing it here puts the connect in the startup
	// number, where every other subject's also is.
	if _, err := io.Read(ctx); err != nil {
		return nil, err
	}
	return io, nil
}

// Read is the contract: one Redis GET, then one parameterised Postgres SELECT.
func (io *Io) Read(ctx context.Context) (IoPayload, error) {
	cached, err := io.redis.Get(ctx, ioRedisKey).Result()
	if err != nil {
		return IoPayload{}, err
	}
	var payload IoPayload
	payload.Cached = cached
	row := io.pool.QueryRow(ctx, ioSelect, ioRowID)
	if err := row.Scan(&payload.ID, &payload.Memo, &payload.Amount); err != nil {
		return IoPayload{}, err
	}
	return payload, nil
}

// WriteTo answers an http.ResponseWriter directly, so cmd/nethttp needs no copy
// of the encode-and-write dance and cmd/gin can hand it gin's writer.
func (io *Io) WriteTo(w http.ResponseWriter, r *http.Request) {
	payload, err := io.Read(r.Context())
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	encoded, err := json.Marshal(payload)
	if err != nil {
		http.Error(w, "", http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(encoded)
}

// IoPoolSize is reported in the subject registry, so the number in the docs and
// the number in the code cannot drift.
var IoPoolSize = strconv.Itoa(ioPoolSize)

File: internal/bench/servers/dotnet/shared/Io.cs
using System.Text.Json.Serialization;
using Npgsql;
using StackExchange.Redis;

namespace Bench;

/// <summary>
/// The io scenario's clients for both .NET subjects: Npgsql with its pool capped
/// at the size every subject is pinned to, and StackExchange.Redis on the one
/// multiplexed connection it is designed around.
///
/// Connected only when the harness passes both URLs, so the other four scenarios
/// run a process that has opened no socket and paid no connect in its startup
/// number.
/// </summary>
public sealed class Io
{
    private const string RedisKey = "bench:greeting";
    private const int RowId = 1;
    private const int PoolSize = 8;
    private const string Select = "SELECT id, memo, amount FROM bench_ledger WHERE id = $1";

    private readonly NpgsqlDataSource? _source;
    private readonly IDatabase? _redis;

    private Io(NpgsqlDataSource? source, IDatabase? redis)
    {
        _source = source;
        _redis = redis;
    }

    /// <summary>
    /// True when the harness enabled the scenario and both clients answered.
    /// `aspnet-minimal` maps the route only then; `aspnet-mvc` declares its action
    /// either way, because MVC reads a controller's routes off the class.
    /// </summary>
    public bool Enabled => _source is not null;

    /// <summary>
    /// Always returns an instance, so the MVC controller can take one by
    /// constructor whether or not the scenario is on. A disabled one throws if
    /// read, and only the io scenario requests that path.
    /// </summary>
    public static async Task<Io> ConnectAsync()
    {
        var pgUrl = Environment.GetEnvironmentVariable("BENCH_IO_PG_URL");
        var redisUrl = Environment.GetEnvironmentVariable("BENCH_IO_REDIS_URL");
        if (string.IsNullOrEmpty(pgUrl) || string.IsNullOrEmpty(redisUrl))
        {
            return new Io(null, null);
        }

        var builder = new NpgsqlDataSourceBuilder(ToNpgsql(pgUrl));
        var source = builder.Build();

        var multiplexer = await ConnectionMultiplexer.ConnectAsync(ToRedisOptions(redisUrl));

        var io = new Io(source, multiplexer.GetDatabase());
        // One round trip here, so the connect lands in the startup number where
        // every other subject's also is.
        await io.ReadAsync();
        return io;
    }

    /// <summary>
    /// StackExchange.Redis does not accept a `redis://` URL - it has its own
    /// comma-separated format - and every subject in the suite is handed the same
    /// URL, so it is taken apart here.
    ///
    /// The port and the credentials are both handled for the reason the Postgres
    /// branch below handles them: a `Uri` with no port reports `-1`, which would
    /// reach `Parse` as `host:-1`, and dropping `UserInfo` would fail against any
    /// Redis that asks for a password. The default URL has neither, which is
    /// exactly why this would have gone unnoticed.
    /// </summary>
    private static ConfigurationOptions ToRedisOptions(string url)
    {
        var parsed = new Uri(url);
        var options = ConfigurationOptions.Parse(
            $"{parsed.Host}:{(parsed.Port == -1 ? 6379 : parsed.Port)}");
        options.AbortOnConnectFail = true;

        var credentials = parsed.UserInfo.Split(':', 2);
        if (credentials.Length > 0 && credentials[0].Length > 0)
        {
            options.User = Uri.UnescapeDataString(credentials[0]);
        }
        if (credentials.Length > 1)
        {
            options.Password = Uri.UnescapeDataString(credentials[1]);
        }

        return options;
    }

    /// <summary>
    /// Npgsql takes a keyword connection string rather than a URL, and every
    /// other subject in the suite is handed the same `postgres://` one.
    /// </summary>
    private static string ToNpgsql(string url)
    {
        var parsed = new Uri(url);
        var credentials = parsed.UserInfo.Split(':', 2);
        return new NpgsqlConnectionStringBuilder
        {
            Host = parsed.Host,
            Port = parsed.Port == -1 ? 5432 : parsed.Port,
            Username = credentials.Length > 0 ? Uri.UnescapeDataString(credentials[0]) : "",
            Password = credentials.Length > 1 ? Uri.UnescapeDataString(credentials[1]) : "",
            Database = parsed.AbsolutePath.TrimStart('/'),
            MaxPoolSize = PoolSize,
            MinPoolSize = 1,
        }.ConnectionString;
    }

    /// <summary>One Redis GET, then one parameterised Postgres SELECT.</summary>
    public async Task<IoPayload> ReadAsync()
    {
        if (_source is null || _redis is null)
        {
            throw new InvalidOperationException("io is not connected: BENCH_IO_PG_URL is unset");
        }

        var cached = await _redis.StringGetAsync(RedisKey);

        await using var command = _source.CreateCommand(Select);
        command.Parameters.AddWithValue(RowId);
        await using var reader = await command.ExecuteReaderAsync();
        if (!await reader.ReadAsync())
        {
            return new IoPayload(cached.ToString() ?? "missing", 0, "missing", 0);
        }

        return new IoPayload(
            cached.ToString() ?? "missing",
            reader.GetInt32(0),
            reader.GetString(1),
            reader.GetInt32(2));
    }
}

/// <summary>
/// Property order is JSON property order for System.Text.Json, and the harness
/// compares bytes.
/// </summary>
public sealed record IoPayload(
    [property: JsonPropertyOrder(0)] string Cached,
    [property: JsonPropertyOrder(1)] int Id,
    [property: JsonPropertyOrder(2)] string Memo,
    [property: JsonPropertyOrder(3)] int Amount);
